WordPress/gutenberg
# Components: refactor `CustomGradientBar` to pass `exhaustive-deps`

## What?
Updates the `CustomGradientBar` component to satisfy the `exhaustive-deps` eslint rule (specifically by updating the `ControlPoints` subcomponent)

## Why?
Part of the effort in https://github.com/WordPress/gutenberg/pull/41166 to apply `exhuastive-deps` to the Components package

## How?

**GradientPickerDropdown**
Update the `popoverProps` `useMemo` dependency array to include `gradientPickerDomRef` instead of `gradientPickerDomRef.current`.

**`useEffect` to clean event listeners on unmount**
The dependency array is currently empty, which works to trigger the cleanup only on unmount. Unfortunately`exhaustive-deps` expects to see the function being called in the dependency array, which would cause it to trigger on every render.

To avoid that, if we wrap `cleanEventListeners` in its own `useCallback` the listeners aren't able to be cleaned properly. Using a ref of the function instead allows us to exclude it from the dependency array so our `useEffect` can provide the intended cleanup, without disrupting the actual user interactions with the component.

Took me a little bit to arrive at this solution, and I'm very open to feedback if there are other/better approaches :) 

## Testing Instructions
1. From your local Gutenberg directory, run `npx eslint --rule 'react-hooks/exhaustive-deps: warn' packages/components/src/custom-gradient-bar`
2. Confirm that the linter returns no errors
3. Confirm e2e tests still pass
